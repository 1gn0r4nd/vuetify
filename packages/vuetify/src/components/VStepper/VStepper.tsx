// Styles
import './VStepper.sass'

// Components
import { VStepperStep } from './VStepperStep'
import { VStepperContent } from './VStepperContent'
import { VStepperLine } from './VStepperLine'
import { VWindow } from '@/components/VWindow'

// Composables
import { makeGroupProps, useGroup } from '@/composables/group'
import { makeTagProps } from '@/composables/tag'

// Utilities
import { computed, provide, toRef } from 'vue'
import { genericComponent } from '@/util'

// Types
import type { InjectionKey, PropType, Ref } from 'vue'
import type { MakeSlots } from '@/util'
import type { GroupProvide } from '@/composables/group'

type VStepperProvide = {
  direction: Ref<string>
  stackedLabels: Ref<boolean | undefined>
}

export const VStepperGroupProvideSymbol: InjectionKey<GroupProvide> = Symbol.for('vuetify:v-stepper-group')
export const VStepperProvideSymbol: InjectionKey<VStepperProvide> = Symbol.for('vuetify:v-stepper')

type VStepperItem<T> = {
  value: T
} | T

function parseItems<T> (items: VStepperItem<T>[] | undefined) {
  if (!items) return []

  return items.map(item => {
    if (typeof item === 'string') return { value: item }

    return item
  })
}

export const VStepper = genericComponent<new <T>() => {
  $props: {
    items?: VStepperItem<T>[]
  }
  $slots: MakeSlots<{
    default: []
  }>
}>()({
  name: 'VStepper',

  props: {
    stackedLabels: Boolean,
    nonLinear: Boolean,
    flat: Boolean,
    direction: {
      type: String,
      default: 'vertical',
      validator: (v: any) => ['vertical', 'horizontal'].includes(v),
    },
    items: Array as PropType<VStepperItem<any>[]>,
    ...makeGroupProps({
      mandatory: 'force' as const,
    }),
    ...makeTagProps(),
  },

  emits: {
    'update:modelValue': (v: any) => true,
  },

  setup (props, { slots, emit }) {
    const group = useGroup(props, VStepperGroupProvideSymbol)
    const items = computed(() => parseItems(props.items))

    provide(VStepperProvideSymbol, {
      stackedLabels: toRef(props, 'stackedLabels'),
      direction: toRef(props, 'direction'),
    })

    const classes = computed(() => ([
      'v-stepper',
      `v-stepper--${props.direction}`,
    ]))

    return () => {
      if (props.direction === 'vertical') {
        return (
          <props.tag class={ classes.value }>
            { items.value.map((item, index) => (
              <>
                <VStepperStep { ...item } step={ index + 1 } v-slots={ slots } />

                <div class="v-stepper__wrapper">
                  <VStepperLine empty={ index + 1 === Number(props.items?.length) } />
                  <VStepperContent value={item.value} step={ index + 1 } v-slots={ slots } />
                </div>
              </>
            )) }
          </props.tag>
        )
      }

      return (
        <props.tag class={ classes.value }>
          <div class="v-stepper__header">
            { items.value.map((item, index) => (
              <>
                <VStepperStep { ...item } step={ index + 1 } v-slots={ slots } />
                { index + 1 < Number(props.items?.length) && (
                  <VStepperLine />
                ) }
              </>
            )) }
          </div>

          <div class="v-stepper__wrapper">
            <VWindow modelValue={ group.selected.value }>
              { props.items?.map((item, index) => (
                <VStepperContent value={item.value} step={ index + 1 } v-slots={ slots } />
              )) }
            </VWindow>
          </div>
        </props.tag>
      )
    }
  },
})
