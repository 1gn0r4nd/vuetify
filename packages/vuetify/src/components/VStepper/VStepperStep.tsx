// Styles
import './VStepperStep.sass'

// Components
import { VAvatar } from '@/components/VAvatar'
import { VBtn } from '@/components/VBtn'

// Composable
import { makeGroupItemProps, useGroupItem } from '@/composables/group'
import { defineComponent } from '@/util'

// Utilities
import { computed, inject } from 'vue'
import { VStepperGroupProvideSymbol, VStepperProvideSymbol } from './VStepper'

// Types
import { useTextColor } from '@/composables/color'

export const VStepperStep = defineComponent({
  name: 'VStepperStep',

  inheritAttrs: false,

  props: {
    color: {
      type: String,
      default: 'primary',
    },
    complete: Boolean,
    completeIcon: {
      type: String,
      default: '$complete',
    },
    editable: Boolean,
    editIcon: {
      type: String,
      default: '$edit',
    },
    errors: Boolean,
    errorIcon: {
      type: String,
      default: '$error',
    },
    // rules: {
    //   type: Array as PropType<any[]>,
    //   default: () => [],
    // },
    step: {
      type: Number,
      required: true,
    },
    title: String,
    subtitle: String,
    ...makeGroupItemProps(),
  },

  setup (props, { slots, emit }) {
    const groupItem = useGroupItem(props, VStepperGroupProvideSymbol)

    const currentStep = computed(() => groupItem.group.items.value.findIndex(x => x.id === groupItem.group.selected.value[0]) + 1)
    const active = computed(() => props.step <= currentStep.value)
    const completed = computed(() => props.step < currentStep.value)

    const icon = computed(() => {
      if (props.errors && props.errorIcon) return props.errorIcon
      if (completed.value && props.completeIcon) return props.completeIcon

      return undefined
    })

    const color = computed(() => {
      if (props.errors) return 'error'
      if (active.value) return 'primary'

      return 'grey'
    })

    const { textColorClasses, textColorStyles } = useTextColor(computed(() => {
      if (props.errors) return 'error'

      return undefined
    }))

    const stepper = inject(VStepperProvideSymbol)

    if (!stepper) throw new Error('foo')

    return () => (
      <VBtn
        class={[
          'v-stepper-step',
          `v-stepper-step--${stepper.direction.value}`,
          {
            'v-stepper-step--stacked-labels': stepper.direction.value === 'horizontal' && stepper.stackedLabels.value,
          },
          textColorClasses.value,
        ]}
        style={{
          ...textColorStyles.value,
        }}
        stacked={ stepper.direction.value === 'horizontal' && stepper.stackedLabels.value }
        variant="text"
        rounded={ 0 }
        onClick={ groupItem.toggle }
        height="100%"
        block={ stepper.direction.value === 'vertical' }
      >
        {{
          prepend: () => (
            <VAvatar
              size="x-small"
              color={ color.value }
              class={ !active.value && 'text-white' }
              icon={ icon.value }
            >
              { !icon.value && props.step }
            </VAvatar>
          ),
          default: () => (
            <div class="v-stepper-step__content">
              <div class="v-stepper-step__label">{ props.title }</div>
              { props.subtitle && <div class="v-stepper-step__subtitle">{ props.subtitle }</div> }
            </div>
          ),
        }}
      </VBtn>
    )
  },
})
