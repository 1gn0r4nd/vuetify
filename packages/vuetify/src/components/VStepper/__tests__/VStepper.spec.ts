// Components
import VStepper from '../VStepper'

// Utilities
import {
  mount,
  MountOptions,
  Wrapper,
} from '@vue/test-utils'

describe('VStepper.ts', () => {
  type Instance = InstanceType<typeof VStepper>
  let mountFunction: (options?: MountOptions<Instance>) => Wrapper<Instance>

  beforeEach(() => {
    mountFunction = (options = {}) => {
      return mount(VStepper, {
        ...options,
      })
    }
  })

  // https://github.com/vuetifyjs/vuetify/issues/10096
  it('should accept 0 as a step value', () => {
    const wrapper = mountFunction({
      propsData: { value: 0 },
    })

    expect(wrapper.vm.internalValue).toBe(0)
  })

  it('should render a flat stepper', () => {
    const wrapper = mountFunction({
      propsData: {
        flat: true,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should have the correct elevation', async () => {
    const wrapper = mountFunction()

    wrapper.setProps({ elevation: 0 })
    expect(wrapper.classes('elevation-0')).toBe(true)

    wrapper.setProps({ elevation: 12 })
    expect(wrapper.classes('elevation-12')).toBe(true)

    wrapper.setProps({ elevation: 24 })
    expect(wrapper.classes('elevation-24')).toBe(true)
  })
})
