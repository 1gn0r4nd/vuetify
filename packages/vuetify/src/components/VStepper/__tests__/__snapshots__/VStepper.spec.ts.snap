// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VStepper.ts should render a flat stepper 1`] = `
<div class="v-stepper v-stepper--flat theme--light">
</div>
`;
